### DISCUSSION TITLE
Heterogenous Graph Predict on All and New Users

### DISCUSSION THREAD

**Discussion by jonimatix** [pyg-team/pytorch_geometric#7020] [Q&A]
Hi,

I am working along side the example for Link Prediction on MovieLens.ipynb (https://colab.research.google.com/drive/1xpzn1Nvai1ygd_P5Yambc_oe4VBPK_ZT?usp=sharing). 

Can you please provide code example on how to predict top N movies for all users, and how to generate recommendations for a completely new user (unseen in training dataset)?

I appreciate your help.

Jon

**Comment by wsad1** [comment]
Update forward to return the embedding
```
    def forward(self, data: HeteroData, return_embedding: False) -> Tensor:
        x_dict = {
          "user": self.user_emb(data["user"].node_id),
          "movie": self.movie_lin(data["movie"].x) + self.movie_emb(data["movie"].node_id),
        } 

        # `x_dict` holds feature matrices of all node types
        # `edge_index_dict` holds all edge indices of all edge types
        x_dict = self.gnn(x_dict, data.edge_index_dict)

        if return_embedding:
            return x_dict
        pred = self.classifier(
            x_dict["user"],
            x_dict["movie"],
            data["user", "rates", "movie"].edge_label_index,
        )

        return pred
```
Then use [knn](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.pool.knn.html#torch_geometric.nn.pool.knn).
```
knn(x_dict['movies'], x_dict['users'], k=N)
```
Also, you can just add completely new users to the graph and pass it to the forward function.

Open to accepting PRs that updates this collab notebook, to show how to predict top N movies for all users.

**Reply by jonimatix** [reply]
Thanks,

So during training loop, do you store the embeddings from each sample?
If so, what do you use as training loss if embeddings are returned?

`
N = 3
for epoch in range(1, 3):
    total_loss = total_examples = 0

    embs = []
    for sampled_data in tqdm(train_loader):
        optimizer.zero_grad()

        sampled_data.to(device)
        emb = model(sampled_data, return_embedding=True)

        embs.append(knn(x=torch.tensor(emb['movie']), y=torch.tensor(emb['user']), k=N))

        optimizer.step()

`

Would be great if you can assist please

Thanks

**Reply by wsad1** [reply]
During training loop you would set `return_embedding=False`. 
I am assuming you need to predict top-k movies only during evalution phase (https://colab.research.google.com/drive/1xpzn1Nvai1ygd_P5Yambc_oe4VBPK_ZT#scrollTo=Yq-I2xaYueF0), in this phase you can set `return_embedding=True`.

**Reply by yuan-meng-dd** [reply]
After collecting embeddings from the mini-batches, how do we find the ID of each node (in this case `MappedID` of movies and `MappedID` of users?) Asking because I want to join each node's embeddings with the input dataframe.

```
# empty list to collect embeddings
embeddings = []

# set model to evaluation mode 
model.eval()

for sampled_data in tqdm.tqdm(val_loader):
    with torch.no_grad():
        # get a batch of data
        sampled_data.to(device)
        
        # node embeddings
        emb = model(sampled_data, return_embedding=True)
        embeddings.append(emb)
```

More specifically, do `sampled_data["movie"]["node_id"]` and  `sampled_data["user"]["node_id"]` return node ID's in the original graph?

**Reply by wsad1** [reply]
Yes `sampled_data["movie"]["node_id"]` should give you original node ids.
